File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/rules_builder/rules_builder_types.py
from dataclasses import dataclass, field
from typing import Any, Dict, List, Union


@dataclass
class InternalRuleDefWithNumericValue:
    value: int


@dataclass
class InternalBase:
    pass


@dataclass
class InternalBaseWithValue:
    value: Any


@dataclass
class InternalBaseWithInt(InternalBaseWithValue):
    value: int


@dataclass
class InternalBaseWithListOfInts(InternalBaseWithValue):
    value: List[int] = field(default_factory=list)

    def __post_init__(self) -> None:
        # copied, so that a rule built from this def can mutate its own value
        self.value = list(self.value)


@dataclass
class InternalRuleDefChar(InternalBaseWithListOfInts):
    pass


@dataclass
class InternalRuleDefCharAlt(InternalBaseWithInt):
    pass


@dataclass
class InternalRuleDefAlt(InternalBase):
    pass


@dataclass
class InternalRuleDefCharNot(InternalBaseWithListOfInts):
    pass


@dataclass
class InternalRuleDefCharRngUpper(InternalBaseWithInt):
    pass


@dataclass
class InternalRuleDefReference(InternalBaseWithInt):
    pass


@dataclass
class InternalRuleDefEnd(InternalBase):
    pass


@dataclass
class InternalRuleDefWithoutValue(InternalBase):
    pass


InternalRuleDef = Union[
    InternalRuleDefChar,
    InternalRuleDefEnd,
    InternalRuleDefReference,
    InternalRuleDefCharNot,
    InternalRuleDefWithNumericValue,
    InternalRuleDefWithoutValue,
    InternalRuleDefAlt,
    InternalRuleDefCharRngUpper,
    InternalRuleDefCharAlt,
]

InternalRuleDefCharOrAltChar = Union[InternalRuleDefChar, InternalRuleDefCharAlt]

SymbolIdsRecord = Dict[str, int]


def is_rule_def_alt(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefAlt)


def is_rule_def_ref(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefReference)


def is_rule_def_end(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefEnd)


def is_rule_def_char(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefChar)


def is_rule_def_char_not(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharNot)


def is_rule_def_char_alt(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharAlt)


def is_rule_def_char_rng_upper(rule: Any) -> bool:
    return isinstance(rule, InternalRuleDefCharRngUpper)

File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/src/rulesBuilder/rulesBuilderTypes.ts
import { repr } from '../utils/repr.ts';

/** Base for the value-less internal rules; exists purely to mirror the reference hierarchy. */
export class InternalBase {
  toRepr(): string {
    return `${this.constructor.name}()`;
  }
}

export class InternalBaseWithValue<T> {
  value: T;

  constructor(value: T) {
    this.value = value;
  }

  toRepr(): string {
    return `${this.constructor.name}(value=${repr(this.value)})`;
  }
}

export class InternalBaseWithInt extends InternalBaseWithValue<number> {}

export class InternalBaseWithListOfInts extends InternalBaseWithValue<number[]> {
  constructor(value: number[] = []) {
    super([...value]);
  }
}

export class InternalRuleDefWithNumericValue extends InternalBaseWithInt {}

export class InternalRuleDefChar extends InternalBaseWithListOfInts {}

export class InternalRuleDefCharAlt extends InternalBaseWithInt {}

export class InternalRuleDefAlt extends InternalBase {}

export class InternalRuleDefCharNot extends InternalBaseWithListOfInts {}

export class InternalRuleDefCharRngUpper extends InternalBaseWithInt {}

export class InternalRuleDefReference extends InternalBaseWithInt {}

export class InternalRuleDefEnd extends InternalBase {}

export class InternalRuleDefWithoutValue extends InternalBase {}

export type InternalRuleDef =
  | InternalRuleDefChar
  | InternalRuleDefEnd
  | InternalRuleDefReference
  | InternalRuleDefCharNot
  | InternalRuleDefWithNumericValue
  | InternalRuleDefWithoutValue
  | InternalRuleDefAlt
  | InternalRuleDefCharRngUpper
  | InternalRuleDefCharAlt;

export type InternalRuleDefCharOrAltChar = InternalRuleDefChar | InternalRuleDefCharAlt;

export const isRuleDefAlt = (rule: unknown): rule is InternalRuleDefAlt =>
  rule instanceof InternalRuleDefAlt;

export const isRuleDefRef = (rule: unknown): rule is InternalRuleDefReference =>
  rule instanceof InternalRuleDefReference;

export const isRuleDefEnd = (rule: unknown): rule is InternalRuleDefEnd =>
  rule instanceof InternalRuleDefEnd;

export const isRuleDefChar = (rule: unknown): rule is InternalRuleDefChar =>
  rule instanceof InternalRuleDefChar;

export const isRuleDefCharNot = (rule: unknown): rule is InternalRuleDefCharNot =>
  rule instanceof InternalRuleDefCharNot;

export const isRuleDefCharAlt = (rule: unknown): rule is InternalRuleDefCharAlt =>
  rule instanceof InternalRuleDefCharAlt;

export const isRuleDefCharRngUpper = (rule: unknown): rule is InternalRuleDefCharRngUpper =>
  rule instanceof InternalRuleDefCharRngUpper;
